README.md
<h1 align="center">Result Type for Dart</h1>

<p align="center">
    <a href="https://github.com/epam-cross-platform-lab/dart_result_type/actions">
    <img src="https://github.com/epam-cross-platform-lab/dart_result_type/workflows/Dart/badge.svg" alt="CI Status" />
  </a>

  <a href="https://codecov.io/gh/epam-cross-platform-lab/dart_result_type">
    <img src="https://codecov.io/gh/epam-cross-platform-lab/dart_result_type/branch/main/graph/badge.svg?token=8HOYKOPG31"/>
  </a>

  <a href="https://github.com/epam-cross-platform-lab/dart_result_type/blob/main/LICENSE">
    <img src="https://img.shields.io/badge/license-Apache-blue.svg" alt="Result Type is released under the Apache license." />
  </a>

  <a href="https://github.com/tenhobi/effective_dart">
    <img src="https://img.shields.io/badge/style-effective_dart-40c4ff.svg" alt="Effective Dart" />
  </a>

  <a href="https://github.com/epam-cross-platform-lab/dart_result_type/blob/main/CODE_OF_CONDUCT.md">
    <img src="https://img.shields.io/badge/PRs-welcome-brightgreen.svg" alt="PRs welcome!" />
  </a>
</p>

# Content

- [Features](#features)
- [Requirements](#requirements)
- [Install](#install)
- [Example](#example)
- [Support](#support)
- [License](#license)

## Features

Result is a type that represents either [Success](https://github.com/epam-cross-platform-lab/dart_result_type/blob/main/lib/src/success.dart) or [Failure](https://github.com/epam-cross-platform-lab/dart_result_type/blob/main/lib/src/failure.dart).

Inspired by [functional programming](http://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Either.html), [Rust](https://doc.rust-lang.org/std/result/enum.Result.html) and [Swift](https://developer.apple.com/documentation/swift/result).

## Requirements

- Dart: 2.12.0+

## Install

```yaml
dependencies:
  result_type: ^0.1.0
```

## Example

The detailed example can be found at [result_type/example/example.dart](https://github.com/epam-cross-platform-lab/dart_result_type/blob/main/example/example.dart).

```dart
import 'dart:async';
import 'dart:convert';
import 'dart:math';

import 'package:http/http.dart' as http;
import 'package:result_type/result_type.dart';

void main() async {
  final random = Random();
  final client = http.Client();
  final result = await getPhotos(client);

  /// Do something with successful operation results or handle an error.
  if (result.isSuccess) {
    print('Photos Items: ${result.success}');
  } else {
    print('Error: ${result.failure}');
  }

  /// Apply transformation to successful operation results or handle an error.
  if (result.isSuccess) {
    final items = result.map((i) => i.where((j) => j.title.length > 60)).success;
    print('Number of Long Titles: ${items.length}');
  } else {
    print('Error: ${result.failure}');
  }

  /// In this example, note the difference in the result of using `map` and
  /// `flatMap` with a transformation that returns an result type.
  Result<int, Error> getNextInteger() => Success(random.nextInt(4));
  Result<int, Error> getNextAfterInteger(int n) => Success(random.nextInt(n + 1));

  final nextIntegerNestedResults = getNextInteger().map(getNextAfterInteger);
  print(nextIntegerNestedResults.runtimeType);
  /// `Prints`: Success<Result<int, Error>, dynamic>

  final nextIntegerUnboxedResults =  getNextInteger().flatMap(getNextAfterInteger);
  print(nextIntegerUnboxedResults.runtimeType);
  /// `Prints`: Success<int, Error>

  /// Use completion handler / callback style API if you want to.
  await getPhotos(client)
    ..result((photos) {
      print('Photos: $photos');
    }, (error) {
      print('Error: $error');
    });
}
```

To see examples of the following package in action:

```sh
cd example && dart run
```

## Support

Post issues and feature requests on the GitHub [issue tracker](https://github.com/epam-cross-platform-lab/dart_result_type/issues).

## License

The source code of Result Type project is available under the Apache license.
See the [LICENSE](https://github.com/epam-cross-platform-lab/dart_result_type/blob/main/LICENSE) file for more info.


CHANGELOG.md
## [0.1.0] - 04.04.2021.

- Migrate to NNBD.

## 0.0.1 - 20.01.2021

 - Initial Release

test/test_all.dart
import 'package:test/test.dart';

import 'failure_test.dart' as failure;
import 'result_test.dart' as result;
import 'success_test.dart' as success;

void main() {
  group('success', success.main);
  group('failure', failure.main);
  group('result', result.main);
}


test/result_test.dart
import 'dart:math';

import 'package:result_type/result_type.dart';
import 'package:test/test.dart';

import 'utils/mock_error.dart';

void main() {
  group('Result:', () {
    Random? random;

    setUp(() {
      random = Random();
    });

    tearDown(() {
      random = null;
    });

    test('Returns Success', () {
      final result = getUser(value: true);
      if (result.isSuccess) {
        print('Success: ${result.success}');
      } else {
        print('Error: ${result.failure}');
      }

      expect(result.success, 'John Doe');
    });

    test('Returns Failure', () {
      final result = getUser(value: false);
      if (result.isSuccess) {
        print('Success: ${result.success}');
      } else {
        print('Error: ${result.failure}');
      }

      expect(result.failure, const MockError(404));
    });

    test('Returns Success From Callbacks', () {
      getUser(value: true)
        ..result((success) {
          expect(success, 'John Doe');
        }, (_) {});
    });

    test('Returns Failure From Callbacks', () {
      getUser(value: false)
        ..result((_) {}, (error) {
          expect(error, const MockError(404));
        });
    });

    test(
        '''Throw Exception when accessing success value without checking if isSuccess''',
        () {
      final result = getUser(value: false);

      expect(() => result.success, throwsException);
    });

    test(
        '''Throw Exception when accessing failure value without checking if isFailure''',
        () {
      final result = getUser(value: true);

      expect(() => result.failure, throwsException);
    });

    test('Apply map transformation to successful operation results', () {
      final result = getUser(value: true);
      final user = result.map((i) => i.toUpperCase()).success;

      expect(user, 'JOHN DOE');
    });

    test(
        '''Throw an error from map transformation without applying transformation to error type''',
        () {
      final result = getUser(value: false);
      final error = result.map((i) => i.toUpperCase()).failure;

      expect(error, const MockError(404));
    });

    test('Apply mapError transformation to failure type', () {
      final error =
          getUser(value: false).mapError((i) => MockError(i.code - 4)).failure;

      expect(error.code, const MockError(400).code);
    });

    test(
        '''Returns successful result without applying mapError transformation''',
        () {
      final maybeError =
          getUser(value: true).mapError((i) => MockError(i.code - 4));

      if (maybeError.isFailure) {
      } else {
        expect(maybeError.success, 'John Doe');
      }
    });

    test('Apply flatMap transformation to successful operation results', () {
      Result<int, MockError> getNextInteger() => Success(random!.nextInt(4));
      Result<int, MockError> getNextAfterInteger(int n) =>
          Success(random!.nextInt(n + 1));

      final nextIntegerUnboxedResults =
          getNextInteger().flatMap(getNextAfterInteger);

      expect(
        nextIntegerUnboxedResults,
        const TypeMatcher<Success<int, MockError>>(),
      );
    });

    test('flatMap does not apply transformation to Failure', () {
      Result<int, MockError> getNextInteger() => Failure(const MockError(451));
      Result<int, MockError> getNextAfterInteger(int n) =>
          Failure(const MockError(404));

      final nextIntegerUnboxedResults =
          getNextInteger().flatMap(getNextAfterInteger);

      expect(
        nextIntegerUnboxedResults,
        const TypeMatcher<Failure<int, MockError>>(),
      );
    });

    test('Apply flatMapError transformation to failure operation results', () {
      Result<int, MockError> getNextInteger() => Failure(const MockError(451));
      Result<int, MockError> getNextAfterInteger(MockError error) =>
          Failure(MockError(error.code));

      final nextIntegerUnboxedResults =
          getNextInteger().flatMapError(getNextAfterInteger);

      expect(
        nextIntegerUnboxedResults,
        const TypeMatcher<Failure<int, MockError>>(),
      );
    });

    test(
        '''flatMapError does not apply transformation to success operation results''',
        () {
      Result<int, MockError> getNextInteger() => Success(random!.nextInt(4));
      Result<int, MockError> getNextAfterInteger(MockError error) =>
          Failure(MockError(error.code));

      final nextIntegerUnboxedResults =
          getNextInteger().flatMapError(getNextAfterInteger);

      expect(
        nextIntegerUnboxedResults,
        const TypeMatcher<Success<int, MockError>>(),
      );
    });
  });
}

Result<String, MockError> getUser({required bool value}) =>
    value ? Success('John Doe') : Failure(const MockError(404));


test/failure_test.dart
import 'package:result_type/src/failure.dart';
import 'package:test/test.dart';

import 'utils/mock_error.dart';

void main() {
  group('Failure', () {
    test('Should return _TestError', () {
      final failure = Failure<int, MockError>(const MockError(1));
      expect(failure.value.code, 1);
    });

    test('Two identical Failures should be equal', () {
      final failure1 = Failure<int, MockError>(const MockError(1));
      final failure2 = Failure<int, MockError>(const MockError(1));

      expect(failure1, failure2);
    });

    test('Two identical Failures should have the same hashCode', () {
      final failure1 = Failure<int, MockError>(const MockError(1));
      final failure2 = Failure<int, MockError>(const MockError(1));

      expect(failure1.hashCode, failure2.hashCode);
    });

    test('Can print to string', () {
      final failure = Failure<int, MockError>(const MockError(1));
      expect(failure.toString(), 'Failure: MockError(code: 1)');
    });
  });
}


lib/src/result.dart
// ignore_for_file: lines_longer_than_80_chars, avoid_shadowing_type_parameters
import 'failure.dart';
import 'success.dart';

/// Callbacks that return [Success] or [Failure].
typedef Completion<T> = void Function(T value);

/// A value that represents either a success or a failure, including an
/// associated value in each case.
abstract class Result<S, F> {
  /// Returns true if [Result] is [Failure].
  bool get isFailure => this is Failure<S, F>;

  /// Returns true if [Result] is [Success].
  bool get isSuccess => this is Success<S, F>;

  /// Returns a new value of [Failure] result.
  ///
  /// Handle an error or do something with successful operation results:
  ///
  /// ```dart
  /// final result = await getPhotos();
  ///
  /// if (result.isFailure) {
  ///   print('Error: ${result.failure}');
  /// } else {
  ///   print('Photos Items: ${result.success}');
  /// }
  /// ```
  ///
  F get failure {
    if (this is Failure<S, F>) {
      return (this as Failure<S, F>).value;
    }

    throw Exception(
      'Make sure that result [isFailure] before accessing [failure]',
    );
  }

  /// Returns a new value of [Success] result.
  ///
  /// Do something with successful operation results or handle an error:
  ///
  /// ```dart
  /// final result = await getPhotos();
  ///
  /// if (result.isSuccess) {
  ///   print('Photos Items: ${result.success}');
  /// } else {
  ///   print('Error: ${result.failure}');
  /// }
  /// ```
  ///
  S get success {
    if (this is Success<S, F>) {
      return (this as Success<S, F>).value;
    }

    throw Exception(
      'Make sure that result [isSuccess] before accessing [success]',
    );
  }

  /// Returns a new value of [Result] from closure
  /// either a success or a failure.
  ///
  /// This example shows how to use completion handler.
  ///
  /// ```dart
  /// final result = await getPhotos();
  ///
  /// await getPhotos(client)
  /// ..result((photos) {
  ///   print('Photos: $photos');
  /// }, (error) {
  ///   print('Error: $error');
  /// });
  /// ```
  ///
  void result(Completion<S> success, Completion<F> failure) {
    if (isSuccess) {
      final left = this as Success<S, F>;
      success(left.value);
    }

    if (isFailure) {
      final right = this as Failure<S, F>;
      failure(right.value);
    }
  }

  /// Maps a [Result<S, F>] to [Result<U, F>] by applying a function
  /// to a contained [Success] value, leaving an [Failure] value untouched.
  /// This function can be used to compose the results of two functions.
  ///
  /// Apply transformation to successful operation results or handle an error:
  ///
  /// ```dart
  /// final result = await getPhotos();
  ///
  /// if (result.isSuccess) {
  ///   final items = result.map((i) => i.where((j) => j.title.length > 60)).success;
  ///   print('Number of Long Titles: ${items.length}');
  /// } else {
  ///   print('Error: ${result.failure}');
  /// }
  /// ```
  ///
  Result<U, F> map<U, F>(U Function(S) transform) {
    if (isSuccess) {
      final left = this as Success<S, F>;
      return Success(transform(left.value));
    } else {
      final right = this as Failure<S, F>;
      return Failure(right.value);
    }
  }

  /// Maps a [Result<S, F>] to [Result<S, E>] by applying a function
  /// to a contained [Failure] value, leaving an [Success] value untouched.
  ///
  /// This function can be used to pass through a successful result
  /// while applying transformation to [Failure].
  ///
  Result<S, E> mapError<S, E>(E Function(F) transform) {
    if (isSuccess) {
      final left = this as Success<S, F>;
      return Success(left.value);
    } else {
      final right = this as Failure<S, F>;
      return Failure(transform(right.value));
    }
  }

  /// Maps a [Result<S, F>] to [Result<U, F>] by applying a function
  /// to a contained [Success] value and unwrapping the produced result,
  /// leaving an [Failure] value untouched.
  ///
  /// Use this method to avoid a nested result when your transformation
  /// produces another [Result] type.
  ///
  /// In this example, note the difference in the result of using `map` and
  /// `flatMap` with a transformation that returns an result type.
  ///
  /// ```dart
  /// Result<int, Error> getNextInteger() => Success(random.nextInt(4));
  /// Result<int, Error> getNextAfterInteger(int n) => Success(random.nextInt(n + 1));
  ///
  /// final nextIntegerNestedResults = getNextInteger().map(getNextAfterInteger);
  /// print(nextIntegerNestedResults.runtimeType);
  /// `Prints: Success<Result<int, Error>, dynamic>`
  ///
  /// final nextIntegerUnboxedResults = getNextInteger().flatMap(getNextAfterInteger);
  /// print(nextIntegerUnboxedResults.runtimeType);
  /// `Prints: Success<int, Error>`
  ///  ```
  Result<U, F> flatMap<U, F>(Result<U, F> Function(S) transform) {
    if (isSuccess) {
      final left = this as Success<S, F>;
      return transform(left.value);
    } else {
      final right = this as Failure<S, F>;
      return Failure(right.value);
    }
  }

  /// Maps a [Result<S, F>] to [Result<S, E>] by applying a function
  /// to a contained [Failure] value, leaving an [Success] value untouched.
  ///
  /// This function can be used to pass through a successful result
  /// while unboxing [Failure] and applying transformation to it.
  ///
  Result<S, E> flatMapError<S, E>(Result<S, E> Function(F) transform) {
    if (isSuccess) {
      final left = this as Success<S, F>;
      return Success(left.value);
    } else {
      final right = this as Failure<S, F>;
      return transform(right.value);
    }
  }
}


example/example.dart
import 'dart:async';
import 'dart:convert';
import 'dart:math';

import 'package:http/http.dart' as http;
import 'package:result_type/result_type.dart';

void main() async {
  final random = Random();
  final client = http.Client();
  final result = await getPhotos(client);

  /// Do something with successful operation results or handle an error.
  if (result.isSuccess) {
    print('Photos Items: ${result.success}');
  } else {
    print('Error: ${result.failure}');
  }

  /// Apply transformation to successful operation results or handle an error.
  if (result.isSuccess) {
    final items =
        result.map((i) => i.where((j) => j.title.length > 60)).success;
    print('Number of Long Titles: ${items.length}');
  } else {
    print('Error: ${result.failure}');
  }

  /// In this example, note the difference in the result of using `map` and
  /// `flatMap` with a transformation that returns an result type.
  Result<int, Error> getNextInteger() => Success(random.nextInt(4));
  Result<int, Error> getNextAfterInteger(int n) =>
      Success(random.nextInt(n + 1));

  final nextIntegerNestedResults = getNextInteger().map(getNextAfterInteger);
  print(nextIntegerNestedResults.runtimeType);
  // Prints: Success<Result<int, Error>, dynamic>

  final nextIntegerUnboxedResults =
      getNextInteger().flatMap(getNextAfterInteger);
  print(nextIntegerUnboxedResults.runtimeType);
  // Prints: Success<int, Error>

  /// Use completion handler / callback style API if you want to.
  await getPhotos(client)
    ..result((photos) {
      print('Photos: $photos');
    }, (error) {
      print('Error: $error');
    });
}

class Photo {
  final int id;
  final String title;
  final String thumbnailUrl;

  const Photo({
    required this.id,
    required this.title,
    required this.thumbnailUrl,
  });

  factory Photo.fromJson(Map<String, Object> json) {
    print(json);
    return Photo(
      id: json['id'] as int,
      title: json['title'] as String,
      thumbnailUrl: json['thumbnailUrl'] as String,
    );
  }

  @override
  String toString() =>
      'Photo(id: $id, title: $title, thumbnailUrl: $thumbnailUrl)';
}

extension PhotoExtension on Photo {
  static List<Photo> parsePhotos(String responseBody) {
    final jsonObject = jsonDecode(responseBody) as Iterable;
    return jsonObject
        .map<Photo>((json) => Photo.fromJson(Map<String, Object>.from(json)))
        .toList();
  }
}

class NetworkError implements Exception {
  final int code;
  final String description;

  const NetworkError({
    required this.code,
    required this.description,
  });

  @override
  String toString() => 'NetworkError(code: $code, description: $description)';

  static const notFound = NetworkError(code: 404, description: 'Not Found');
}

Future<Result<List<Photo>, NetworkError>> getPhotos(http.Client client) async {
  const path = 'https://jsonplaceholder.typicode.com/photos';
  try {
    final jsonString = await client.get(Uri.parse(path));
    return Success(PhotoExtension.parsePhotos(jsonString.body));
  } on NetworkError catch (_) {
    return Failure(NetworkError.notFound);
  }
}
